Dockerfile

```diff
@@ -1,160 +1,186 @@
-    ARCH=amd64 \
-    PYTHON_VERSION=3.6 \
-    PATH=$HOME/.local/bin/:$PATH \
+# syntax=docker/dockerfile:1.2
+# Stream swaps to CentOS Stream once and is reused.
+FROM docker.io/library/centos:8 AS stream
+RUN set -ex\
+	; dnf -y -q install centos-release-stream\
+	; dnf -y -q swap centos-{linux,stream}-repos\
+# This "|| true" is needed if building in podman, it seems.
+# The filesystem package tries to modify permissions on /proc.
+	; dnf -y -q distro-sync || true\
+    ; dnf -y -q clean all
+
+# Base is set up with the runtime dependencies and environment.
+FROM stream AS base
+# Only set variables or install packages that need to end up in the
+# final container here.
+ENV PATH=/app/bin/:$PATH \
     PYTHONUNBUFFERED=1 \
     PYTHONIOENCODING=UTF-8 \
     LC_ALL=C.UTF-8 \
     LANG=C.UTF-8
-
 ENV QUAYDIR /quay-registry
 ENV QUAYCONF /quay-registry/conf
-ENV QUAYPATH "."
+ENV QUAYPATH $QUAYDIR
+ENV PYTHONUSERBASE /app
+ENV PYTHONPATH $QUAYPATH
+RUN set -ex\
+	; dnf -y -q --setopt=tsflags=nodocs --setopt=skip_missing_names_on_install=False install\
+		dnsmasq \
+		memcached \
+		nginx \
+		openldap \
+		openssl \
+		python3 \
+		python3-gpg \
+		skopeo \
+     ; dnf -y -q clean all
+
+# Build has all the build-only tools.
+FROM base AS build
+ENV PYTHONDONTWRITEBYTECODE 1
+RUN set -ex\
+	; dnf -y -q --setopt=tsflags=nodocs --setopt=skip_missing_names_on_install=False install\
+		gcc-c++\
+		git\
+		nodejs\
+		openldap-devel\
+		python3-devel\
+    ; dnf -y -q clean all
+RUN alternatives --set python /usr/bin/python3
+WORKDIR /build
+
+# Config-editor builds the javascript for the configtool.
+FROM build AS config-editor
+# This argument must be repeated, and should have the same default as
+# the other CONFIGTOOL_VERSION argument.
+ARG CONFIGTOOL_VERSION=master
+RUN curl -fsSL "https://github.com/quay/config-tool/archive/${CONFIGTOOL_VERSION}.tar.gz"\
+	| tar xz --strip-components=4 --exclude=static --exclude='*.go'\
+		'*/pkg/lib/editor'
+RUN set -ex\
+	; npm install --quiet --no-progress --ignore-engines\
+    ; npm run --quiet build\
+	;
+
+# Build-python installs the requirements for the python code.
+FROM build AS build-python
+COPY requirements.txt .
+# Note that it installs into PYTHONUSERBASE because of the '--user'
+# flag.
+RUN set -ex\
+	; python3 -m pip install --no-cache-dir --quiet\
+		--upgrade setuptools pip\
+    ; python3 -m pip install --no-cache-dir --progress-bar off\
+		--user --requirement requirements.txt --no-cache\
+	;
+RUN set -ex\
+# Doing this is explicitly against the purpose and use of certifi.
+	; for dir in\
+		$(find "$(python3 -m site --user-base)" -type d -name certifi)\
+	; do chgrp -R 0 "$dir" && chmod -R g=u "$dir" ; done\
+	;
+
+# Build-static downloads the static javascript.
+FROM build-python AS build-static
+# The external library versions rarely change, do them first.
+COPY external_libraries.py _init.py .
+RUN set -ex\
+	; mkdir -p static/{webfonts,fonts,ldn}\
+    ; python3 -m external_libraries\
```

Comment:
`external_libraries.py` depends on some Quay module that I think are missing in this particular build container

Reply:
There's a patch earlier in branch to fix that.